### PR TITLE
Improve opencl

### PR DESCRIPTION
**Pull request by A2va**
* Before adding new features and new modules, please go to issues to submit the relevant feature description first.
* Write good commit messages and use the same coding conventions as the rest of the project.
* Please commit code to dev branch and we will merge into master branch in feature
* Ensure your edited codes with four spaces instead of TAB.

------

* 增加新特性和新模块之前，请先到issues提交相关特性说明，经过讨论评估确认后，再进行相应的代码提交，避免做无用工作。
* 编写友好可读的提交信息，并使用与工程代码相同的代码规范，代码请用4个空格字符代替tab缩进。
* 请提交代码到dev分支，如果通过，我们会在特定时间合并到master分支上。
* 为了规范化提交日志的格式，commit消息，不要用中文，请用英文描述。

### CODE REVIEW

**Comment by A2va** [comment]
All the windows builds are failing because it seems that xmake didn't add `-c core.longpaths=true` to git args. I leave the logs here:

```
git clone https://github.com/KhronosGroup/OpenCL-SDK.git -c core.fsmonitor=false source.tmp\opencl
Cloning into 'source.tmp\opencl'...
git -c core.fsmonitor=false checkout ae7fcae82fe0b7bcc272e43fc324181b2d544eea
Note: switching to 'ae7fcae82fe0b7bcc272e43fc324181b2d544eea'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at ae7fcae :Reduce warning in OpenCL SDK (#74)
git -c core.fsmonitor=false submodule update --init --recursive
Submodule 'headers-cpp' (https://github.com/KhronosGroup/OpenCL-CLHPP.git) registered for path 'external/OpenCL-CLHPP'
Submodule 'headers' (https://github.com/KhronosGroup/OpenCL-Headers.git) registered for path 'external/OpenCL-Headers'
Submodule 'loader' (https://github.com/KhronosGroup/OpenCL-ICD-Loader.git) registered for path 'external/OpenCL-ICD-Loader'
Cloning into 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-CLHPP'...
Cloning into 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-Headers'...
Cloning into 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-ICD-Loader'...
Submodule path 'external/OpenCL-CLHPP': checked out '8f536659b8f98220938afdd6b6cf6c90dc03b615'
Submodule 'external/CMock' (https://github.com/ThrowTheSwitch/CMock) registered for path 'external/OpenCL-CLHPP/external/CMock'
Submodule 'external/Unity' (https://github.com/ThrowTheSwitch/Unity) registered for path 'external/OpenCL-CLHPP/external/Unity'
Cloning into 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-CLHPP/external/CMock'...
Cloning into 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-CLHPP/external/Unity'...
Submodule path 'external/OpenCL-CLHPP/external/CMock': checked out '7cc41ddfdd07dc5eb8359d278f439f14031d64ad'
Submodule 'vendor/c_exception' (https://github.com/throwtheswitch/cexception.git) registered for path 'external/OpenCL-CLHPP/external/CMock/vendor/c_exception'
Submodule 'vendor/unity' (https://github.com/throwtheswitch/unity.git) registered for path 'external/OpenCL-CLHPP/external/CMock/vendor/unity'
Cloning into 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-CLHPP/external/CMock/vendor/c_exception'...
Cloning into 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-CLHPP/external/CMock/vendor/unity'...
Submodule path 'external/OpenCL-CLHPP/external/CMock/vendor/c_exception': checked out 'dce9e8b26f2179439002e02d691429e81a32b6c0'
Submodule 'vendor/unity' (https://github.com/throwtheswitch/unity.git) registered for path 'external/OpenCL-CLHPP/external/CMock/vendor/c_exception/vendor/unity'
Cloning into 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-CLHPP/external/CMock/vendor/c_exception/vendor/unity'...
fatal: cannot write keep file 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/.git/modules/headers-cpp/modules/external/CMock/modules/vendor/c_exception/modules/vendor/unity/objects/pack/pack-b57fbca2c381e963a6526ad441c177c991522dda.keep': Filename too long
fatal: fetch-pack: invalid index-pack output
fatal: clone of 'https://github.com/throwtheswitch/unity.git' into submodule path 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-CLHPP/external/CMock/vendor/c_exception/vendor/unity' failed
Failed to clone 'vendor/unity'. Retry scheduled
Cloning into 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-CLHPP/external/CMock/vendor/c_exception/vendor/unity'...
fatal: cannot write keep file 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/.git/modules/headers-cpp/modules/external/CMock/modules/vendor/c_exception/modules/vendor/unity/objects/pack/pack-e4fbce862a91a8a513063d888920446cf2c4a078.keep': Filename too long
fatal: fetch-pack: invalid index-pack output
fatal: clone of 'https://github.com/throwtheswitch/unity.git' into submodule path 'C:/Users/runneradmin/AppData/Local/.xmake/cache/packages/2306/o/opencl/v2023.04.17/source.tmp/opencl/external/OpenCL-CLHPP/external/CMock/vendor/c_exception/vendor/unity' failed
Failed to clone 'vendor/unity' a second time, aborting
fatal: Failed to recurse into submodule path 'external/OpenCL-CLHPP/external/CMock/vendor/c_exception'
Submodule path 'external/OpenCL-CLHPP/external/CMock/vendor/unity': checked out '031f3bbe45f8adf504ca3d13e6f093869920b091'
fatal: Failed to recurse into submodule path 'external/OpenCL-CLHPP/external/CMock'
Submodule path 'external/OpenCL-CLHPP/external/Unity': checked out '7d2bf62b7e6afaf38153041a9d53c21aeeca9a25'
fatal: Failed to recurse into submodule path 'external/OpenCL-CLHPP'
Submodule path 'external/OpenCL-Headers': checked out '4fdcfb0ae675f2f63a9add9552e0af62c2b4ed30'
Submodule path 'external/OpenCL-ICD-Loader': checked out 'b1bce7c3c580a8345205cf65fc1a5f55ba9cdb01'
error: @programdir\core\sandbox\modules\os.lua:378: execv(git -c core.fsmonitor=false submodule update --init --recursive) failed(1)
````

**Comment by A2va** [comment]
Ok, this time another problem. The problem is that there is a symlink in the github repository that makes it impossible to copy (Previously I tried using the archive download, but because of that 7zip couldn't extract archive.). The only option I see to solve this is to use the prebuilts provided by the project, unfortunately it only exists for x64 or x86, arm would then be unsupported.
```
error: @programdir\modules\private\async\runjobs.lua:256: @programdir\core\sandbox\modules\os.lua:100: cannot copy directory C:\Users\runneradmin\AppData\Local\.xmake\cache\packages\2306\o\opencl\v2023.04.17\source\opencl to C:\Users\RUNNER~1\AppData\Local\Temp\.xmake\230629.dir,  Unknown
 ```
What do you think ? Is it worth opening an issue to the project to tell them about this problem? At least remove the symlink from the releases source archive, it's impossible to unzip it with 7zip because of that.

File: packages/o/opencl/xmake.lua [hunk 1]
```diff
@@ -64,3 +69,9 @@ package("opencl")
             end
         end
     end)
+
+    on_install("windows", "linux","macosx","android",function (package)
```
Comment (waruqi) [inline packages/o/opencl/xmake.lua:73]: add space, ` , `

**Comment by waruqi** [comment]
> Ok, this time another problem. The problem is that there is a symlink in the github repository that makes it impossible to copy (Previously I tried using the archive download, but because of that 7zip couldn't extract archive.). The only option I see to solve this is to use the prebuilts provided by the project, unfortunately it only exists for x64 or x86, arm would then be unsupported.
> 
> ```
> error: @programdir\modules\private\async\runjobs.lua:256: @programdir\core\sandbox\modules\os.lua:100: cannot copy directory C:\Users\runneradmin\AppData\Local\.xmake\cache\packages\2306\o\opencl\v2023.04.17\source\opencl to C:\Users\RUNNER~1\AppData\Local\Temp\.xmake\230629.dir,  Unknown
> ```
> 
> What do you think ? Is it worth opening an issue to the project to tell them about this problem? At least remove the symlink from the releases source archive, it's impossible to unzip it with 7zip because of that.

you can using  prebuilts binary first.

File: packages/o/opencl/xmake.lua [hunk 2]
```diff
@@ -64,3 +77,15 @@ package("opencl")
...
+    on_install("windows|x86", "windows|x64", function (package)
+        os.cp("lib", package:installdir())
+        os.cp("bin", package:installdir())
+        os.cp("include", package:installdir())
+    end)
+
+    on_install("linux", "macosx", "android", function (package)
+        local configs = {"-DOPENCL_SDK_BUILD_SAMPLES=OFF"}
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
+        import("package.tools.cmake").install(package, configs)
```
Comment (waruqi) [inline packages/o/opencl/xmake.lua:90]: > tools/cmake.lua:979: cmake not found!


you need add `add_deps("cmake")`

**Pull request merged**